org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java

```diff
@@ -246,64 +249,48 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer) {
 
         Types.Capabilities caps = new Types.Capabilities();
         caps.supportsConfigurationDoneRequest = true;
-        caps.supportsDelayedStackTraceLoading = true;
+        caps.supportsRestartRequest = true;
+        caps.supportTerminateDebuggee = true;
         return new Responses.InitializeResponseBody(caps);
     }
 
     private Responses.ResponseBody launch(Requests.LaunchArguments arguments) {
-        this.cwd = arguments.cwd;
-        String mainClass = arguments.startupClass;
-        String classpath;
+        // Need cache the launch json because VSCode doesn't resend the launch json at the RestartRequest.
+        this.launchArguments = arguments;
         try {
-            classpath = AdapterUtils.computeClassPath(arguments.projectName, mainClass);
-            classpath = classpath.replaceAll("\\\\", "/");
-        } catch (CoreException e) {
-            Logger.logException("Failed to resolve classpath.", e);
-            return new Responses.ErrorResponseBody(new Types.Message(3001, "Cannot launch jvm.", null));
-        }
-
-        if (arguments.sourcePath == null || arguments.sourcePath.length == 0) {
-            this.sourcePath = new String[] { cwd };
-        } else {
-            this.sourcePath = new String[arguments.sourcePath.length];
-            System.arraycopy(arguments.sourcePath, 0, this.sourcePath, 0, arguments.sourcePath.length);
+            this.launchDebugSession(arguments);
+        } catch (DebugException e) {
+            return new Responses.ErrorResponseBody(new Types.Message(3001, "Cannot launch debuggee vm: " + e.getMessage(), null));
```

Comment:
Accord to debugProtocol, in ErrorResponse body, we could set an optional structure Message. And the Message has a required id field. The code 3001 just reference to OpenDebugAD7 error code.